pkg/jobstore/boltdb/store.go

```diff
@@ -0,0 +1,1013 @@
+			err = bkt.ForEach(func(key []byte, data []byte) error {
+				var item model.JobHistory
+
+				err := json.Unmarshal(data, &item)
+				if err != nil {
+					return err
+				}
+
+				history = append(history, item)
+				return nil
+			})
+
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if !options.ExcludeExecutionLevel {
+		// 	// Get the executions for this JobID
+		if bkt, err := NewBucketPath(BucketPathExecutionHistory, jobID).Get(tx, false); err != nil {
+			return nil, err
+		} else {
+			err = bkt.ForEach(func(key []byte, data []byte) error {
+				var item model.JobHistory
+
+				err := json.Unmarshal(data, &item)
+				if err != nil {
+					return err
+				}
+
+				history = append(history, item)
+				return nil
+			})
+
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// Filter out anything before the specified Since time
+	history = lo.Filter(history, func(item model.JobHistory, index int) bool {
+		return item.Time.Unix() >= options.Since
+	})
+
+	sort.Slice(history, func(i, j int) bool { return history[i].Time.UTC().Before(history[j].Time.UTC()) })
+
+	return history, nil
+}
+
+// CreateJob creates a new record of a job in the data store
+func (b *BoltJobStore) CreateJob(ctx context.Context, job model.Job) error {
+	return b.database.Update(func(tx *bolt.Tx) (err error) {
+		return b.createJob(tx, job)
+	})
+}
+
+func (b *BoltJobStore) createJob(tx *bolt.Tx, job model.Job) error {
+	if b.jobExists(tx, job.ID()) {
+		return jobstore.NewErrJobAlreadyExists(job.Metadata.ID)
+	}
+
+	jobIDKey := []byte(job.Metadata.ID)
+
+	jobState := model.JobState{
+		JobID:      job.Metadata.ID,
+		State:      model.JobStateNew,
+		Version:    1,
+		CreateTime: b.clock.Now().UTC(),
+		UpdateTime: b.clock.Now().UTC(),
+	}
+	data, err := json.Marshal(jobState)
+	if err != nil {
+		return err
+	}
+
+	// Write the JobState to the state bucket
+	if bkt, err := NewBucketPath(BucketPathState).Get(tx, false); err != nil {
+		return err
+	} else {
+		if err = bkt.Put(jobIDKey, data); err != nil {
+			return err
+		}
+	}
+
+	// Write the job to the Job bucket
+	jobData, err := json.Marshal(job)
+	if err != nil {
+		return err
+	}
+	if bkt, err := NewBucketPath(BucketPathJobs).Get(tx, false); err != nil {
+		return err
+	} else {
+		if err = bkt.Put(jobIDKey, jobData); err != nil {
+			return err
+		}
+	}
+
+	if err = b.inProgressIndex.Add(tx, BucketJobsInProgress, jobIDKey); err != nil {
+		return err
+	}
+
+	if job.Metadata.ClientID == "" {
+		fmt.Println("TEST missing clientID in job")
+		return errors.New("job is missing a client id")
+	}
+
+	if err = b.clientsIndex.Add(tx, []byte(job.Metadata.ClientID), jobIDKey); err != nil {
+		return err
+	}
+
+	// Write sentinels keys for specific tags
+	for _, tag := range job.Spec.Annotations {
+		tagBytes := []byte(strings.ToLower(tag))
+		if err = b.tagsIndex.Add(tx, tagBytes, jobIDKey); err != nil {
+			return err
+		}
+	}
+
+	b.triggerEvent(jobstore.JobWatcher, jobstore.CreateEvent, jobData)
```

Comment:
the events are triggered within the transaction, which is great as it guarantees ordering of events published to match the order of mutations and updates to the underlying store. However, a slow event consumer can block the transaction, and you might see a deadlock if the event consumer tries to read data from the datastore while processing the event.

I think the transactions should just publish the events to a channel and have another go routine read from the channel and notify the watchers. Does that make sense? Though I don't want to add complexity to this PR and we can do it in a follow up PR if you prefer

Reply:
I might attempt this in another PR if that's okay.  

If we take this approach, won't the new go routine reading from the channel block if any of the called watchers block when handling the notification?   Or do we need a new goroutine per notification and hope that the recipient returns at some point?